Drafted fix

maryamariyan opened the pull request:
cc: @janvorli, @bartonjs

Proper fix for compilation issue caused by deprecated API in macOS Mojave by

using dlsym to call available API rather than suppressing deprecation warnings.

Fixes: #30599

maryamariyan commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509.c:
```diff
@@ -52,7 +53,19 @@ AppleCryptoNative_X509GetPublicKey(SecCertificateRef cert, SecKeyRef* pPublicKey
+    // and SecCertificateCopyPublicKey on the systems that don’t.
+    SecKeyRef (*secCertificateCopyKey)(SecCertificateRef);
+    OSStatus (*secCertificateCopyPublicKey)(SecCertificateRef, SecKeyRef*);
+    if ((secCertificateCopyKey = (SecKeyRef (*)(SecCertificateRef))dlsym(RTLD_DEFAULT, "SecCertificateCopyKey")) != NULL)
```
@janvorli is there a way we could call dlsym only once in this code snippet?

maryamariyan commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509.c:
```diff
@@ -52,7 +53,27 @@ AppleCryptoNative_X509GetPublicKey(SecCertificateRef cert, SecKeyRef* pPublicKey
+    
+    if (!checked)
+    {
+        secCertificateCopyKey = (SecKeyRef (*)(SecCertificateRef))dlsym(RTLD_DEFAULT, "SecCertificateCopyKey");
```
@janvorli is there a way we could call dlsym only once in this code snippet?

bartonjs commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509.c:
```diff
@@ -52,7 +53,27 @@ AppleCryptoNative_X509GetPublicKey(SecCertificateRef cert, SecKeyRef* pPublicKey
+    {
+        secCertificateCopyKey = (SecKeyRef (*)(SecCertificateRef))dlsym(RTLD_DEFAULT, "SecCertificateCopyKey");
+        secCertificateCopyPublicKey = (OSStatus (*)(SecCertificateRef, SecKeyRef*))dlsym(RTLD_DEFAULT, "SecCertificateCopyPublicKey");
+        checked = 1;
```
In C# I'd do this as `Volatile.Write(ref checked, 1)`, or declare `checked` as `volatile`.  Is declaring `checked` `volatile` here the right way to prevent a reordering race condition, @janvorli?

bartonjs replied:
(Or is my `static local`-fu weak, and we don't need to do a lazy initializer manually, because static locals guarantee that they only run their initializer once? (and therefore the two dlsym calls should move up to variable initializers))

bartonjs commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509.h:
```diff
@@ -42,7 +42,7 @@ Returns 1 on success, 0 on failure, any other value on invalid state.
 Output:
 pPublicKeyOut: Receives a CFRetain()ed SecKeyRef for the public key
-pOSStatusOut: Receives the result of SecCertificateCopyPublicKey
+pOSStatusOut: Receives the result of SecCertificateCopyKey
```
Receives the result of SecCertificateCopyKey or SecCertificateCopyPublicKey, depending on the OS version.

maryamariyan committed:
Proper fix for compilation issue caused by deprecated API in macOS Mojave by
using dlsym to call available API rather than suppressing deprecation warnings.

Fixes: #30599

maryamariyan commented:
Closing this PR and opening https://github.com/dotnet/corefx/pull/30815 on corefx fork instead
